docs/http-cloud-provider.md

```diff
@@ -0,0 +1,643 @@
+
+##### Example Config File:
+
+```bash
+[Global]
+client-url = http://127.0.0.1 
+```
+
+First the API checks for which interfaces in [cloud.go](../pkg/cloudprovider/cloud.go) are implemented.
+
+A GET request for /v1/{Interface}/supported will return whether the interface is supported or not.
+
+The following methods must be implemented by the HTTP cloudprovider. If a cloudprovider does not implement a resource, it can return false  or return a HTTP 404 Error with empty body for a given GET /{interface}/supported call.
+
+If you encounter an error in processing the request, return an HTTP 404 with a response body of:
+
+```json
+{
+    "error" :"error-message-here"
+}
+```
+
+## Provider Name
+
+#### GET /v1/providerName
+
+Returns the cloud provider name.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|providerName|Name of the cloud provider.|"my-cloud"|
+Example
+
+```json
+{
+    "providerName": "my-cloud"
+}
+```
+
+## Instances
+
+The instances interface allows the kubernetes platform to query the cloud provider about the instances currently within the cloud provider system.
+
+### Methods
+
+#### GET /v1/instances/supported
+
+Returns whether the Instance interface is supported or not.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|supported|Whether the interface is supported or not.|true|
+Example
+
+```json
+{
+    "supported": true
+}
+```
+
+#### GET /v1/instances/{instanceName}/nodeAddresses
+
+Returns all node addresses for the given instance.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|nodeAddresses|List of node adress objects.||
+|nodeAddresses:type|The IP type relate to the node address. The types of addresses are restricted to the list below.|"InternalIP"|
+|nodeAddresses:address|The actual node's address.|"127.0.0.1"|
+
+Address Types
+
+|Name|Description|
+|----|-----------|
+|InternalIP|The node's Internal IP address.|
+|ExternalIP|The node's External IP address.|
+|Hostname|The hostname of the node.|
+|LegacyHostIP|The node's Host IP address.|
+Example
+
+```json
+{
+    "nodeAddresses": [
+        {
+            "type": "InternalIP",
+            "address": "127.0.0.1"
+        }
+    ]
+}
+```
+
+#### GET /v1/instances/{instanceName}/ID
+
+Returns the Instance ID for the given instance.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|instanceID|ID of the cloud provider.|"my-cloud-name1"|
+Example
+
+```json
+{
+    "instanceID": "my-cloud-name1"
+}
+```
+
+#### GET /v1/instances/{filter}
+
+Returns all instances where the name matches the word filter. filter is a regular expression.
```

Comment:
What type of regular expression? A go regular expression? Are implementers going to have to port golang flavor regexp to whatever language they are writing the cloudprovider in?